cmake/modules/AddSwift.cmake

```diff
@@ -956,9 +956,16 @@ function(add_swift_fuzzer_host_tool executable)
   target_link_libraries(${executable} PRIVATE "-fsanitize=fuzzer")
 endfunction()
 
-macro(add_swift_tool_symlink name dest component)
+macro(add_swift_tool_symlink name dest component destination)
+  # This is used by llvm_install_symlink to determine where
+  # to put the target link
+  # we decide to define this "locally" to the macro instead of the
+  # main CMakeLists.txt so to align with how we configure installation
+  # across our scripts (i.e. granular control for each program in a component)
+  set(SWIFT_TOOLS_INSTALL_DIR destination)
   llvm_add_tool_symlink(SWIFT ${name} ${dest} ALWAYS_GENERATE)
   llvm_install_symlink(SWIFT ${name} ${dest} ALWAYS_GENERATE COMPONENT ${component})
+  unset(SWIFT_TOOLS_INSTALL_DIR)
```

Comment:
Instead of creating an LLVM component called "compiler", can you instead use the default behavior of LLVM (which should create install-${name} as an install command) and instead use something like the following to register that llvm component (swift-frontend) with the swift component (compiler).

function (add_dependency_from_llvm_component_to_swift_component swift-target-name swift-component)
  add_dependencies(install-${swift-target-name} install-${swift-component})
endfunction()

Reply:
From my experiments, it seems I cannot drop the `COMPONENT` argument -- if I do, the `cmake_install.cmake` script will not install the symlink when specifying our Swift components (e.g. `-DCMAKE_INSTALL_COMPONENT=compiler`), but when specifying the name of the target program (if `ALWAYS_GENERATE` is provided) or the source one (otherwise), which is not something that we currently allow in the Swift build system.

As discussed offline, I've attempted to lean a bit more on the LLVM functions around components by (a) dropping `ALWAYS_GENERATE` and (b) configuring an LLVM component for `swift-frontend` to avoid failures -- you can find this in 87f89066740 (although I have mixed feelings about it)